### PR TITLE
[Merged by Bors] - UCB: preparing for proposal/ballot validation

### PR DESCRIPTION
**Pull request by countvonzero**
## Motivation

2nd PR for #2902

## Changes

- needs smesher to specify LayerIndex in Ballot so validation can easily verify whether the ballot declares the right layer index
- use gomock for mocking system.Fetcher for tests in activation/blocks packages
- move utility functions shared btwn block oracle and eligibility validator to proposals package.
  the proposals package will eventually replace blocks package after #2902 is implemented. at the end of UCB implementation, blocks package will be re-created for the UCB data.
  
## Test Plan

unit tests

## DevOps Notes

- [x] This PR does not require configuration changes (e.g., environment variables, GitHub secrets, VM resources)
- [x] This PR does not affect public APIs
- [x] This PR does not rely on a new version of external services (PoET, elasticsearch, etc.)
- [x] This PR does not make changes to log messages (which monitoring infrastructure may rely on)

### CODE REVIEW

File: blocks/blockoracle_test.go [hunk 1]
```diff
@@ -182,7 +182,7 @@ func TestBlockOracleEmptyActiveSet(t *testing.T) {
 	blockOracle := NewMinerBlockOracle(committeeSize, layersPerEpoch, activationDB, beaconProvider, vrfsgn, nodeID, func() bool { return true }, lg.WithName("blockOracle"))
 
 	_, proofs, _, err := blockOracle.BlockEligible(types.NewLayerID(layersPerEpoch * 3))
-	r.EqualError(err, "zero total weight not allowed")
+	r.EqualError(err, "oracle get num slots: zero total weight not allowed")
```
Comment (dshulyak) [inline blocks/blockoracle_test.go:186]: would be nicer to use ErrorIs for this checks

File: common/types/ballot.go [hunk 2]
```diff
@@ -106,16 +110,37 @@ type VotingEligibilityProof struct {
...
+func (b *Ballot) Initialize() error {
+	if b.ID() != EmptyBallotID {
+		return fmt.Errorf("ballot already initialized")
+	}
+
+	bytes := b.Bytes()
+	b.ballotID = BallotID(CalcHash32(bytes).ToHash20())
+	pubkey, err := ed25519.ExtractPublicKey(bytes, b.Signature)
+	if err != nil {
+		return fmt.Errorf("ballot initialize: %w", err)
```
Comment (dshulyak) [inline common/types/ballot.go]: should be about extracting pubkey from sig?

**Review by dshulyak — APPROVED** [review]

File: common/types/ballot.go (hunk 2 above)
Comment (countvonzero) [reply common/types/ballot.go]: done

File: blocks/blockoracle_test.go (hunk 1 above)
Comment (countvonzero) [reply blocks/blockoracle_test.go:186]: done

**Comment by countvonzero** [comment]
bors merge

**Pull request closed without merging**